negative float declaration panic

The following program `sample.go` triggers a panic:

```go
package main
  
func main() {
    a := -1.2
    println(a)
}
```

Expected result:
```console
$ go run ./sample.go
-1.200000e+000
```

Got:
```console
$ yaegi ./sample.go
panic: reflect: call of reflect.Value.Int on float64 Value [recovered]
	panic: reflect: call of reflect.Value.Int on float64 Value

goroutine 1 [running]:
github.com/containous/yaegi/interp.runCfg.func1(0xc0001d2140)
	/Users/marc/go/src/github.com/containous/yaegi/interp/run.go:106 +0x134
panic(0x1829e80, 0xc0001602e0)
	/usr/local/Cellar/go/1.12.5/libexec/src/runtime/panic.go:522 +0x1b5
reflect.Value.Int(...)
	/usr/local/Cellar/go/1.12.5/libexec/src/reflect/value.go:962
github.com/containous/yaegi/interp.negate.func1(0xc0001d2140, 0x1935958)
	/Users/marc/go/src/github.com/containous/yaegi/interp/run.go:1103 +0x15f
github.com/containous/yaegi/interp.runCfg(0xc0001dca00, 0xc0001d2140)
	/Users/marc/go/src/github.com/containous/yaegi/interp/run.go:111 +0x62
github.com/containous/yaegi/interp.(*Interpreter).run(0xc0001441e0, 0xc0001dc200, 0xc00016e050)
	/Users/marc/go/src/github.com/containous/yaegi/interp/run.go:93 +0x126
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc0001441e0, 0xc0001d8000, 0x35, 0x235, 0xc0001d8000, 0x35, 0x0, 0xc0001b8270)
	/Users/marc/go/src/github.com/containous/yaegi/interp/interp.go:284 +0x32d
main.main()
	/Users/marc/go/src/github.com/containous/yaegi/yaegi.go:59 +0x3ec
```

--- a/_test/op3.go
+++ b/_test/op3.go
@@ -4,8 +4,11 @@ import "fmt"
 
 func main() {
 	a := -1.2
-	fmt.Println(a)
+	fmt.Printf("a: %v %T\n", a, a)
+	b := -(2 + 1i)
+	fmt.Printf("b: %v %T\n", b, b)
 }
 
 // Output:
-// -1.2
+// a: -1.2 float64
+// b: (-2-1i) complex128

--- a/cmd/genop/genop.go
+++ b/cmd/genop/genop.go
@@ -149,14 +149,14 @@ func {{$name}}(n *Node) {
 	case reflect.Complex64, reflect.Complex128:
 		switch {
 		case c0.rval.IsValid():
-			r0 := c0.rval.Complex()
+			r0 := vComplex(c0.rval)
 			v1 := genValue(c1)
 			n.exec = func(f *Frame) Builtin {
 				dest(f).SetComplex(r0 {{$op.Name}} v1(f).Complex())
 				return next
 			}
 		case c1.rval.IsValid():
-			r1 := c1.rval.Complex()
+			r1 := vComplex(c1.rval)
 			v0 := genValue(c0)
 			n.exec = func(f *Frame) Builtin {
 				dest(f).SetComplex(v0(f).Complex() {{$op.Name}} r1)
@@ -224,7 +224,7 @@ func {{$name}}Assign(n *Node) {
 			}
 		case reflect.Complex64, reflect.Complex128:
 			v0 := genValue(c0)
-			v1 := c1.rval.Complex()
+			v1 := vComplex(c1.rval)
 			n.exec = func(f *Frame) Builtin {
 				v := v0(f)
 				v.SetComplex(v.Complex() {{$op.Name}} v1)
@@ -401,6 +401,77 @@ func {{$name}}(n *Node) {
 				}
 			}
 		}
+	{{- if $op.Complex}}
+	case isComplex(t0) || isComplex(t1):
+		switch {
+		case c0.rval.IsValid():
+			s0 := vComplex(c0.rval)
+			v1 := genValueComplex(c1)
+			if n.fnext != nil {
+				fnext := getExec(n.fnext)
+				n.exec = func(f *Frame) Builtin {
+					_, s1 := v1(f)
+					if s0 {{$op.Name}} s1 {
+						dest(f).SetBool(true)
+						return tnext
+					}
+					dest(f).SetBool(false)
+					return fnext
+				}
+			} else {
+				n.exec = func(f *Frame) Builtin {
+					_, s1 := v1(f)
+					dest(f).SetBool(s0 {{$op.Name}} s1)
+					return tnext
+				}
+			}
+		case c1.rval.IsValid():
+			s1 := vComplex(c1.rval)
+			v0 := genValueComplex(c0)
+			if n.fnext != nil {
+				fnext := getExec(n.fnext)
+				n.exec = func(f *Frame) Builtin {
+					_, s0 := v0(f)
+					if s0 {{$op.Name}} s1 {
+						dest(f).SetBool(true)
+						return tnext
+					}
+					dest(f).SetBool(false)
+					return fnext
+				}
+			} else {
+				dest := genValue(n)
+				n.exec = func(f *Frame) Builtin {
+					_, s0 := v0(f)
+					dest(f).SetBool(s0 {{$op.Name}} s1)
+					return tnext
+				}
+			}
+		default:
+			v0 := genValueComplex(n.child[0])
+			v1 := genValueComplex(n.child[1])
+			if n.fnext != nil {
+				fnext := getExec(n.fnext)
+				n.exec = func(f *Frame) Builtin {
+					_, s0 := v0(f)
+					_, s1 := v1(f)
+					if s0 {{$op.Name}} s1 {
+						dest(f).SetBool(true)
+						return tnext
+					}
+					dest(f).SetBool(false)
+					return fnext
+				}
+			} else {
+				n.exec = func(f *Frame) Builtin {
+					_, s0 := v0(f)
+					_, s1 := v1(f)
+					dest(f).SetBool(s0 {{$op.Name}} s1)
+					return tnext
+				}
+			}
+		}
+		{{- end}}
 	case isFloat(t0) || isFloat(t1):
 		switch {
 		case c0.rval.IsValid():
@@ -620,10 +691,11 @@ func {{$name}}(n *Node) {
 
 // Op define operator name and properties
 type Op struct {
-	Name  string // +, -, ...
-	Str   bool   // true if operator applies to string
-	Float bool   // true if operator applies to float
-	Shift bool   // true if operator is a shift operation
+	Name    string // +, -, ...
+	Str     bool   // true if operator applies to string
+	Float   bool   // true if operator applies to float
+	Complex bool   // true if operator applies to complex
+	Shift   bool   // true if operator is a shift operation
 }
 
 func main() {
@@ -636,29 +708,29 @@ func main() {
 	b := &bytes.Buffer{}
 	data := map[string]interface{}{
 		"Arithmetic": map[string]Op{
-			"add":    {"+", true, true, false},
-			"sub":    {"-", false, true, false},
-			"mul":    {"*", false, true, false},
-			"quo":    {"/", false, true, false},
-			"rem":    {"%", false, false, false},
-			"shl":    {"<<", false, false, true},
-			"shr":    {">>", false, false, true},
-			"and":    {"&", false, false, false},
-			"or":     {"|", false, false, false},
-			"xor":    {"^", false, false, false},
-			"andnot": {"&^", false, false, false},
+			"add":    {"+", true, true, true, false},
+			"sub":    {"-", false, true, true, false},
+			"mul":    {"*", false, true, true, false},
+			"quo":    {"/", false, true, true, false},
+			"rem":    {"%", false, false, false, false},
+			"shl":    {"<<", false, false, false, true},
+			"shr":    {">>", false, false, false, true},
+			"and":    {"&", false, false, false, false},
+			"or":     {"|", false, false, false, false},
+			"xor":    {"^", false, false, false, false},
+			"andnot": {"&^", false, false, false, false},
 		},
 		"IncDec": map[string]Op{
 			"inc": {Name: "+"},
 			"dec": {Name: "-"},
 		},
 		"Comparison": map[string]Op{
-			"equal":        {Name: "=="},
-			"greater":      {Name: ">"},
-			"greaterEqual": {Name: ">="},
-			"lower":        {Name: "<"},
-			"lowerEqual":   {Name: "<="},
-			"notEqual":     {Name: "!="},
+			"equal":        {Name: "==", Complex: true},
+			"greater":      {Name: ">", Complex: false},
+			"greaterEqual": {Name: ">=", Complex: false},
+			"lower":        {Name: "<", Complex: false},
+			"lowerEqual":   {Name: "<=", Complex: false},
+			"notEqual":     {Name: "!=", Complex: true},
 		},
 	}
 	if err = parse.Execute(b, data); err != nil {

--- a/interp/interp.go
+++ b/interp/interp.go
@@ -195,7 +195,6 @@ func initUniverse() *Scope {
 		"println": &Symbol{kind: Bltn, builtin: _println},
 		"real":    &Symbol{kind: Bltn, builtin: _real},
 		"recover": &Symbol{kind: Bltn, builtin: _recover},
-		// TODO: complex, copy, delete, imag, new, print, real
 	}}
 	return scope
 }

--- a/interp/op.go
+++ b/interp/op.go
@@ -124,14 +124,14 @@ func add(n *Node) {
 	case reflect.Complex64, reflect.Complex128:
 		switch {
 		case c0.rval.IsValid():
-			r0 := c0.rval.Complex()
+			r0 := vComplex(c0.rval)
 			v1 := genValue(c1)
 			n.exec = func(f *Frame) Builtin {
 				dest(f).SetComplex(r0 + v1(f).Complex())
 				return next
 			}
 		case c1.rval.IsValid():
-			r1 := c1.rval.Complex()
+			r1 := vComplex(c1.rval)
 			v0 := genValue(c0)
 			n.exec = func(f *Frame) Builtin {
 				dest(f).SetComplex(v0(f).Complex() + r1)
@@ -374,14 +374,14 @@ func mul(n *Node) {
 	case reflect.Complex64, reflect.Complex128:
 		switch {
 		case c0.rval.IsValid():
-			r0 := c0.rval.Complex()
+			r0 := vComplex(c0.rval)
 			v1 := genValue(c1)
 			n.exec = func(f *Frame) Builtin {
 				dest(f).SetComplex(r0 * v1(f).Complex())
 				return next
 			}
 		case c1.rval.IsValid():
-			r1 := c1.rval.Complex()
+			r1 := vComplex(c1.rval)
 			v0 := genValue(c0)
 			n.exec = func(f *Frame) Builtin {
 				dest(f).SetComplex(v0(f).Complex() * r1)
@@ -558,14 +558,14 @@ func quo(n *Node) {
 	case reflect.Complex64, reflect.Complex128:
 		switch {
 		case c0.rval.IsValid():
-			r0 := c0.rval.Complex()
+			r0 := vComplex(c0.rval)
 			v1 := genValue(c1)
 			n.exec = func(f *Frame) Builtin {
 				dest(f).SetComplex(r0 / v1(f).Complex())
 				return next
 			}
 		case c1.rval.IsValid():
-			r1 := c1.rval.Complex()
+			r1 := vComplex(c1.rval)
 			v0 := genValue(c0)
 			n.exec = func(f *Frame) Builtin {
 				dest(f).SetComplex(v0(f).Complex() / r1)
@@ -874,14 +874,14 @@ func sub(n *Node) {
 	case reflect.Complex64, reflect.Complex128:
 		switch {
 		case c0.rval.IsValid():
-			r0 := c0.rval.Complex()
+			r0 := vComplex(c0.rval)
 			v1 := genValue(c1)
 			n.exec = func(f *Frame) Builtin {
 				dest(f).SetComplex(r0 - v1(f).Complex())
 				return next
 			}
 		case c1.rval.IsValid():
-			r1 := c1.rval.Complex()
+			r1 := vComplex(c1.rval)
 			v0 := genValue(c0)
 			n.exec = func(f *Frame) Builtin {
 				dest(f).SetComplex(v0(f).Complex() - r1)
@@ -1007,7 +1007,7 @@ func addAssign(n *Node) {
 			}
 		case reflect.Complex64, reflect.Complex128:
 			v0 := genValue(c0)
-			v1 := c1.rval.Complex()
+			v1 := vComplex(c1.rval)
 			n.exec = func(f *Frame) Builtin {
 				v := v0(f)
 				v.SetComplex(v.Complex() + v1)
@@ -1192,7 +1192,7 @@ func mulAssign(n *Node) {
 			}
 		case reflect.Complex64, reflect.Complex128:
 			v0 := genValue(c0)
-			v1 := c1.rval.Complex()
+			v1 := vComplex(c1.rval)
 			n.exec = func(f *Frame) Builtin {
 				v := v0(f)
 				v.SetComplex(v.Complex() * v1)
@@ -1321,7 +1321,7 @@ func quoAssign(n *Node) {
 			}
 		case reflect.Complex64, reflect.Complex128:
 			v0 := genValue(c0)
-			v1 := c1.rval.Complex()
+			v1 := vComplex(c1.rval)
 			n.exec = func(f *Frame) Builtin {
 				v := v0(f)
 				v.SetComplex(v.Complex() / v1)
@@ -1546,7 +1546,7 @@ func subAssign(n *Node) {
 			}
 		case reflect.Complex64, reflect.Complex128:
 			v0 := genValue(c0)
-			v1 := c1.rval.Complex()
+			v1 := vComplex(c1.rval)
 			n.exec = func(f *Frame) Builtin {
 				v := v0(f)
 				v.SetComplex(v.Complex() - v1)
@@ -1767,6 +1767,75 @@ func equal(n *Node) {
 		default:
 			v0 := genValueString(n.child[0])
 			v1 := genValueString(n.child[1])
+			if n.fnext != nil {
+				fnext := getExec(n.fnext)
+				n.exec = func(f *Frame) Builtin {
+					_, s0 := v0(f)
+					_, s1 := v1(f)
+					if s0 == s1 {
+						dest(f).SetBool(true)
+						return tnext
+					}
+					dest(f).SetBool(false)
+					return fnext
+				}
+			} else {
+				n.exec = func(f *Frame) Builtin {
+					_, s0 := v0(f)
+					_, s1 := v1(f)
+					dest(f).SetBool(s0 == s1)
+					return tnext
+				}
+			}
+		}
+	case isComplex(t0) || isComplex(t1):
+		switch {
+		case c0.rval.IsValid():
+			s0 := vComplex(c0.rval)
+			v1 := genValueComplex(c1)
+			if n.fnext != nil {
+				fnext := getExec(n.fnext)
+				n.exec = func(f *Frame) Builtin {
+					_, s1 := v1(f)
+					if s0 == s1 {
+						dest(f).SetBool(true)
+						return tnext
+					}
+					dest(f).SetBool(false)
+					return fnext
+				}
+			} else {
+				n.exec = func(f *Frame) Builtin {
+					_, s1 := v1(f)
+					dest(f).SetBool(s0 == s1)
+					return tnext
+				}
+			}
+		case c1.rval.IsValid():
+			s1 := vComplex(c1.rval)
+			v0 := genValueComplex(c0)
+			if n.fnext != nil {
+				fnext := getExec(n.fnext)
+				n.exec = func(f *Frame) Builtin {
+					_, s0 := v0(f)
+					if s0 == s1 {
+						dest(f).SetBool(true)
+						return tnext
+					}
+					dest(f).SetBool(false)
+					return fnext
+				}
+			} else {
+				dest := genValue(n)
+				n.exec = func(f *Frame) Builtin {
+					_, s0 := v0(f)
+					dest(f).SetBool(s0 == s1)
+					return tnext
+				}
+			}
+		default:
+			v0 := genValueComplex(n.child[0])
+			v1 := genValueComplex(n.child[1])
 			if n.fnext != nil {
 				fnext := getExec(n.fnext)
 				n.exec = func(f *Frame) Builtin {
@@ -3212,6 +3281,75 @@ func notEqual(n *Node) {
 		default:
 			v0 := genValueString(n.child[0])
 			v1 := genValueString(n.child[1])
+			if n.fnext != nil {
+				fnext := getExec(n.fnext)
+				n.exec = func(f *Frame) Builtin {
+					_, s0 := v0(f)
+					_, s1 := v1(f)
+					if s0 != s1 {
+						dest(f).SetBool(true)
+						return tnext
+					}
+					dest(f).SetBool(false)
+					return fnext
+				}
+			} else {
+				n.exec = func(f *Frame) Builtin {
+					_, s0 := v0(f)
+					_, s1 := v1(f)
+					dest(f).SetBool(s0 != s1)
+					return tnext
+				}
+			}
+		}
+	case isComplex(t0) || isComplex(t1):
+		switch {
+		case c0.rval.IsValid():
+			s0 := vComplex(c0.rval)
+			v1 := genValueComplex(c1)
+			if n.fnext != nil {
+				fnext := getExec(n.fnext)
+				n.exec = func(f *Frame) Builtin {
+					_, s1 := v1(f)
+					if s0 != s1 {
+						dest(f).SetBool(true)
+						return tnext
+					}
+					dest(f).SetBool(false)
+					return fnext
+				}
+			} else {
+				n.exec = func(f *Frame) Builtin {
+					_, s1 := v1(f)
+					dest(f).SetBool(s0 != s1)
+					return tnext
+				}
+			}
+		case c1.rval.IsValid():
+			s1 := vComplex(c1.rval)
+			v0 := genValueComplex(c0)
+			if n.fnext != nil {
+				fnext := getExec(n.fnext)
+				n.exec = func(f *Frame) Builtin {
+					_, s0 := v0(f)
+					if s0 != s1 {
+						dest(f).SetBool(true)
+						return tnext
+					}
+					dest(f).SetBool(false)
+					return fnext
+				}
+			} else {
+				dest := genValue(n)
+				n.exec = func(f *Frame) Builtin {
+					_, s0 := v0(f)
+					dest(f).SetBool(s0 != s1)
+					return tnext
+				}
+			}
+		default:
+			v0 := genValueComplex(n.child[0])
+			v1 := genValueComplex(n.child[1])
 			if n.fnext != nil {
 				fnext := getExec(n.fnext)
 				n.exec = func(f *Frame) Builtin {

--- a/interp/run.go
+++ b/interp/run.go
@@ -1110,6 +1110,11 @@ func negate(n *Node) {
 			dest(f).SetFloat(-value(f).Float())
 			return next
 		}
+	case reflect.Complex64, reflect.Complex128:
+		n.exec = func(f *Frame) Builtin {
+			dest(f).SetComplex(-value(f).Complex())
+			return next
+		}
 	}
 }
 

--- a/interp/type.go
+++ b/interp/type.go
@@ -171,6 +171,13 @@ func nodeType(interp *Interpreter, scope *Scope, n *Node) (*Type, error) {
 			t.cat = ByteT
 			t.name = "byte"
 			t.untyped = true
+		case complex64:
+			t.cat = Complex64T
+			t.name = "complex64"
+		case complex128:
+			t.cat = Complex128T
+			t.name = "complex128"
+			t.untyped = true
 		case float32:
 			t.cat = Float32T
 			t.name = "float32"
@@ -699,6 +706,14 @@ func isUint(t *Type) bool {
 	return false
 }
 
+func isComplex(t *Type) bool {
+	switch t.TypeOf().Kind() {
+	case reflect.Complex64, reflect.Complex128:
+		return true
+	}
+	return false
+}
+
 func isFloat(t *Type) bool {
 	switch t.TypeOf().Kind() {
 	case reflect.Float32, reflect.Float64:
@@ -715,6 +730,6 @@ func isByteArray(t *Type) bool {
 
 func isFloat32(t *Type) bool       { return t.TypeOf().Kind() == reflect.Float32 }
 func isFloat64(t *Type) bool       { return t.TypeOf().Kind() == reflect.Float64 }
-func isUntypedNumber(t *Type) bool { return t.untyped && (isInt(t) || isFloat(t)) }
-func isNumber(t *Type) bool        { return isInt(t) || isFloat(t) }
+func isUntypedNumber(t *Type) bool { return t.untyped && (isInt(t) || isFloat(t) || isComplex(t)) }
+func isNumber(t *Type) bool        { return isInt(t) || isFloat(t) || isComplex(t) }
 func isString(t *Type) bool        { return t.TypeOf().Kind() == reflect.String }

--- a/interp/value.go
+++ b/interp/value.go
@@ -140,6 +140,20 @@ func vUint(v reflect.Value) (i uint64) {
 	return
 }
 
+func vComplex(v reflect.Value) (c complex128) {
+	switch v.Type().Kind() {
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		c = complex(float64(v.Int()), 0)
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
+		c = complex(float64(v.Uint()), 0)
+	case reflect.Float32, reflect.Float64:
+		c = complex(v.Float(), 0)
+	case reflect.Complex64, reflect.Complex128:
+		c = v.Complex()
+	}
+	return
+}
+
 func vFloat(v reflect.Value) (i float64) {
 	switch v.Type().Kind() {
 	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
@@ -190,6 +204,22 @@ func genValueFloat(n *Node) func(*Frame) (reflect.Value, float64) {
 		return func(f *Frame) (reflect.Value, float64) { v := value(f); return v, float64(v.Uint()) }
 	case reflect.Float32, reflect.Float64:
 		return func(f *Frame) (reflect.Value, float64) { v := value(f); return v, v.Float() }
+	}
+	return nil
+}
+
+func genValueComplex(n *Node) func(*Frame) (reflect.Value, complex128) {
+	value := genValue(n)
+
+	switch n.typ.TypeOf().Kind() {
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return func(f *Frame) (reflect.Value, complex128) { v := value(f); return v, complex(float64(v.Int()), 0) }
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
+		return func(f *Frame) (reflect.Value, complex128) { v := value(f); return v, complex(float64(v.Uint()), 0) }
+	case reflect.Float32, reflect.Float64:
+		return func(f *Frame) (reflect.Value, complex128) { v := value(f); return v, complex(v.Float(), 0) }
+	case reflect.Complex64, reflect.Complex128:
+		return func(f *Frame) (reflect.Value, complex128) { v := value(f); return v, v.Complex() }
 	}
 	return nil
 }

